### DISCUSSION TITLE
Add extension events to hooks

### DISCUSSION THREAD

**Discussion by Attacler** [directus/directus#20306] [Feature Requests]
### Summary

Hi all,

In one of the more recent versions of Directus, the option to enable/disable extensions has been introduced.
This is great but in some cases you need to know when the extension is turned on or off.
For that reason i propose to add 2 new action events: extension.load and extension.unload.
This way you can run code to for example clean up data in Redis or files that it creates to use on the machine.

### Basic Example
```ts
action('extension.load', async (context) => {
          console.log("We are loaded :)");
});

action('extension.unload', async (context) => {
          console.log("We are unloaded :)");
});
```
### Motivation

Right now we have the events server.start and server.stop.
However these events are not being triggerd whenever an extension loads.
For example if the marketplace is introduced and someone installs an extension, the extension does not receive an event for the load, or when someone uninstalls the extension, there is no way of knowing that.

### Detailed Design

I think the previouse explainations make sense, otherwise please let me know.

### Must Have:
- New action event: extension.load
- New action event: extension.unload

### Should Have:
- n/a
### Could Have:
- n/a
### Won't Have:
- Trigger support for flows

### Drawbacks

The drawback of this is that people could use this to start tracking the amount of (un)installs or something similar.

### Alternatives

For load, you can also add your code to the code:
export default ({ action }, { env, services }) => {};
However, this will not capture the unload/disable of an extension.

### Adoption Strategy

This is extra functionality so this should not break any systems.
It should be usable the same way that other action hooks can be used.

### Unresolved Questions

One thing im not sure of is what parameters this should receive apart from the context object.

**Comment by w0kyj** [comment]
Possible workaround with Flows or Hooks:
Enable trigger on the directus_extensions collection - create/update/delete as needed and then send notifications, emails, alerts as needed.

Also noted in Request Review call:
In theory this could already be implemented by the plugins themselves with the extensions Emitter

**Comment by u12206050** [comment]
Events are ok for other extensions/code to be alerted about an extension being added/removed, but I think the cleanest could be adding a `onInstall`, `onEnable` and `onDisable` options to the extension options. This way extensions have a dedicated place for themselves to take actions such as running custom migrations, installing additional extensions(dependencies) etc.

**Comment by Dominic-Marcelino** [comment]
Also related https://github.com/directus/directus/discussions/22095 (less detailed description, but same goal). 

I agree to @u12206050 that `onInstall`, `onEnable` `onDisable` and also an `onUninstall` would be great hooks.

`onUninstall` would be useful that extensions could clean up behind them when being removed. A problem would be to decide if / how they can still listen to it, in case the've already been disabled

**Comment by Nitwel** [comment]
This has been added in https://github.com/directus/directus/pull/25181
